### ISSUE TITLE
Validator error messages are misleading or confusing.

### ISSUE THREAD

**Issue by devjack** [Mermade/swagger2openapi#24]
Schema validation displays "expected $ref" instead of more helpful "Unexpected field `foo` expected a $ref or Schema."

### Checklist

- [ ] Validation: I believe my source definition is invalid OpenAPI 3.0.x but the validator does not complain

_Or rather, "I thought my definition was valid and the error message was unhelpful"._

### Detailed Description
openapi.yml:
```
...
components:
    schemas:
        SomeObject:
            $ref: 'resources/myobject.yml#/resource/SomeObject'
...
```

resources/myobject.yml
```
resource:
    SomeObject:
      name: SomeObject #Intentionally invalid
      type: object #Intentionally Invalid
      properties:
        Id:
          type: string
          format: uuid
          example: "d2014f64-ffdf-487b-8d12-67a20976aca6"
```
In this example the `name` and `type` fields are not part of a [SchemaObject](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#schemaObject) and so the validator should logically fail.

`components/schema` is expected to be  a: `Map[string, Schema Object | Reference Object]` (see [componentsObject](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#componentsObject).

In this scenario the validator fails with:

```
{
    "keyword": "required",
    "dataPath": "/components/schemas/SomeObject",
    "schemaPath": "#/required",
    "params": {
      "missingProperty": "$ref"
    },
    "message": "should have required property '$ref'"
...
```

This is misleading because SomeObject should not have a property $ref; instead, the $ref in `components/schema` pointed to an invalid schema.

### Proposed solution:

A better message should either be:

> Object at /components/schemas/SomeObject is not a valid schemaObject

or

> `$ref: 'resources/myobject.yml#/resource/SomeObject'` points to an invalid schema object, schema object expected.

... or some variant.

**Comment by MikeRalphson** [comment]
Agreed. We rely on the OASv3 schema (which is a work-in-progress) to catch errors the hand-written validator doesn't catch, in this case I felt writing a JSON-schema Draft 5 validator was too much work to duplicate, so the error comes from `ajv`.

The unintuitive error message arises because the definition of the `schemaObject` does not allow `additionalProperties` so it assumes that your object must be a `referenceObject` instead - as this does allow `additionalProperties` (because the spec. says they must be ignored) . Hence the `missing $ref` error.

Do you get an outer error, or a reference to a parent schema saying something like `schema does not match anyOf...`? In such cases the whole error needs to be parsed to find the correct context.

**Comment by MikeRalphson** [comment]
Working on this. How about:

> ../openapi3-examples/fail/schemaProperties.yaml
#/components/schemas/SomeObject
Schema object cannot have additionalProperty: name

`type` of course being a [valid property](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#properties) of the OpenAPI3 `schemaObject`.

**Comment by MikeRalphson** [comment]
Should be fixed in v2.10.0 - please chase if not.
